gensim/models/sent2vec_inner.pyx

```diff
@@ -0,0 +1,117 @@
+     REAL_t, EXP_TABLE, \
+     our_dot, our_saxpy, \
+     our_dot_double, our_dot_float, our_dot_noblas, our_saxpy_noblas
+
+from word2vec import FAST_VERSION
+
+cdef int ONE = 1
+cdef float ONEF = <float>1.0
+
+
+cdef (float, int) negative_sampling(const int target, const float lr, float *grad, float *wo, float *hidden,
+                              const int vector_size, const int neg, int *negatives,
+                              int negpos, const int negatives_len):
+
+    cdef float loss
+    loss = <float> 0.0
+    cdef int label_true, label_false
+    label_true = <int> 1
+    label_false = <int> 0
+    cdef int temp, new_negpos
+    for i from 0 <= i <= neg:
+        if i == 0:
+            loss += binary_logistic(target, label_true, lr, vector_size, wo, grad, hidden)
+        else:
+            temp, new_negpos = get_negative(target, negatives, negpos, negatives_len)
+            loss += binary_logistic(temp, label_false, lr, vector_size, wo, grad, hidden)
+    return loss, new_negpos
+
+
+cdef float binary_logistic(const int target, const int label, const float lr,
+                            const int vector_size, float *wo, float *grad, float *hidden):
+
+    cdef float score = exp(our_dot(&vector_size, &wo[<int>target], &ONE, hidden, &ONE))
+    cdef float alpha = lr * (<float>label - score)
+    our_saxpy(&vector_size, &alpha, &wo[<int>target], &ONE, grad, &ONE)
+    our_saxpy(&vector_size, &alpha, hidden, &ONE, &wo[<int>target], &ONE)
+    if label is True:
+        #return -LOG_TABLE[<int>score]
+        return -log(score)
+    else:
+        #return -LOG_TABLE[<int>(ONEF - score)]
+        return -log(ONEF - score)
+    
+
+cdef (int, int) get_negative(const int target, int *negatives,
+                              int negpos, const int negatives_len):
+
+    cdef int negative
+    while True:
+        negative = negatives[<int>negpos]
+        negpos = (negpos + <int>ONE) % negatives_len
+        if target != negative:
+            break
+    return negative, negpos
+
+
+def update(model, context_, target_, lr_):
+
+    cdef float loss_
+    cdef float lr = <float> lr_
+    cdef int vector_size = <int> model.vector_size
+    cdef int *context = <int *>np.PyArray_DATA(context_)
+    cdef int target = <int> target_
+    cdef int *negatives = <int *>np.PyArray_DATA(model.negatives)
+    cdef int negpos = <int>model.negpos
+    cdef int neg = <int> model.neg
+    cdef REAL_t *wi = <REAL_t *>np.PyArray_DATA(model.wi)
+    cdef REAL_t *wo = <REAL_t *>np.PyArray_DATA(model.wo)
+    cdef int negatives_len = <int>len(model.negatives)
+    cdef int context_len = <int>len(context_)
+
+    #hidden_ = zeros(vector_size, dtype=REAL)
+    #grad_ = zeros(vector_size, dtype=REAL)
+    #cdef float *hidden = <float *>np.PyArray_DATA(hidden_)
+    #cdef float *grad = <float *>np.PyArray_DATA(grad_)
+    cdef REAL_t *hidden, *grad
+    hidden = <REAL_t *>calloc(vector_size, cython.sizeof(REAL_t))
```

Comment:
Dynamic allocations are very expensive (because of book keeping, but also because they ruin caches). Can't we pre-allocate the structures somehow?
(here and elsewhere)